src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1062,14 +1062,29 @@ private AdjustmentRule GetAdjustmentRuleForTime(DateTime dateTime, bool dateTime
                 (dateTime + BaseUtcOffset).Date :
                 dateTime.Date;
 
-            for (int i = 0; i < _adjustmentRules.Length; i++)
+            int low = 0;
+            int high = _adjustmentRules.Length - 1;
+
+            while (low <= high)
             {
-                AdjustmentRule rule = _adjustmentRules[i];
-                AdjustmentRule previousRule = i > 0 ? _adjustmentRules[i - 1] : rule;
-                if (IsAdjustmentRuleValid(rule, previousRule, dateTime, date, dateTimeisUtc))
+                int median = low + ((high - low) >> 1);
+
+                AdjustmentRule rule = _adjustmentRules[median];
+                AdjustmentRule previousRule = median > 0 ? _adjustmentRules[median - 1] : rule;
+
+                int compareResult = CompareAdjustmentRule(rule, previousRule, dateTime, date, dateTimeisUtc);
```

Comment:
>CompareAdjustmentRule [](start = 36, length = 21)

nit: the name CompareAdjustmentRule is kind of confusing that I read it comparing 2 adjustment rules together.